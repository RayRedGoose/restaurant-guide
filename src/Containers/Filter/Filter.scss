.filter {
  ul {
    height: 5rem;
    overflow-x: scroll;
  }
}
