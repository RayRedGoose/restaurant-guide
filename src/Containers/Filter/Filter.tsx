import "./Filter.scss";
import React, { useState, useEffect, ChangeEvent } from "react";
import { useSelector, useDispatch } from "react-redux";
import states from "assets/ts/states";
import { IAppStore } from "assets/ts/interfaces";
import {
  addStateFilter,
  removeStateFilter,
  addGenreFilter,
  removeGenreFilter,
  addAttireFilter,
  removeAttireFilter,
} from "redux_utils/actions";
import { handleFocus, handleBlur } from "./events";

interface Props {
  type: propType;
}

interface Lists {
  states: string[];
  genres: string[];
  attires: string[];
}

type propType = "states" | "genres" | "attires";

const Filter: React.FC<Props> = ({ type }) => {
  const dispatch = useDispatch();
  const initialText: string = "all " + type;
  const [active, setActive] = useState<string>(initialText);
  const {
    genres,
    attires,
    attireFilter,
    genreFilter,
    stateFilter,
  } = useSelector((store: IAppStore) => ({
    genres: store.genres,
    attires: store.attires,
    attireFilter: store.attireFilter,
    genreFilter: store.genreFilter,
    stateFilter: store.stateFilter,
  }));

  const lists: Lists = {
    states: states,
    genres: genres,
    attires: attires,
  };

  const filters = {
    states: stateFilter,
    genres: genreFilter,
    attires: attireFilter,
  };

  const typeFilter: string = filters[type];

  useEffect(() => {
    if (typeFilter === "") setActive(initialText);
  }, [typeFilter, initialText]);

  const chooseAddingCreator = (filter: string): void => {
    if (type === "states") dispatch(addStateFilter(filter));
    if (type === "genres") dispatch(addGenreFilter(filter));
    if (type === "attires") dispatch(addAttireFilter(filter));
  };

  const removeFilter = (): void => {
    if (type === "states") dispatch(removeStateFilter());
    if (type === "genres") dispatch(removeGenreFilter());
    if (type === "attires") dispatch(removeAttireFilter());
  };

  const toggleFilter = (e: ChangeEvent<HTMLSelectElement>) => {
    e.target.size = 1;
    setActive(e.target.value);

    return e.target.value !== initialText
      ? chooseAddingCreator(e.target.value)
      : removeFilter();
  };

  const lines: JSX.Element[] = lists[type].map((li: string) => (
    <option key={li}>{li}</option>
  ));

  return (
    <section className="filter">
      <select
        onChange={toggleFilter}
        onFocus={handleFocus}
        onBlur={handleBlur}
        value={active}
      >
        <option>{"all " + type}</option>
        {lines}
      </select>
    </section>
  );
};

export default Filter;
