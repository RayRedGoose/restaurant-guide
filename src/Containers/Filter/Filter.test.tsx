import React from 'react';
import { shallow } from 'enzyme';
import Filter from './Filter';
import {
  addStateFilter, removeStateFilter,
  addGenreFilter, removeGenreFilter,
  addAttireFilter, removeAttireFilter
} from 'redux_utils/actions';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
  useSelector: () => ({
    genres: ['american'],
    attires: ['casual']
  })
}));

jest.mock('redux_utils/actions');

describe("Filter", () => {
  let wrapper: any;

  it('should match the snapshot if type is states', () => {
    wrapper = shallow(<Filter type="states" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is genres', () => {
    wrapper = shallow(<Filter type="genres" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is attires', () => {
    wrapper = shallow(<Filter type="attires" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("Click Events", () => {
    beforeEach(() => {
      wrapper = shallow(<Filter type="genres" />);
    });

    it('should show unordered list if title is clicked', () => {
      let ul: JSX.Element = wrapper.find('ul');
      expect(ul).toHaveLength(0);
      wrapper.find('h3').simulate('click');

      ul = wrapper.find('ul');
      expect(ul).toHaveLength(1);
    });

    it('should hide opened unordered list if title is clicked again', () => {
      wrapper.find('h3').simulate('click');
      let ul: JSX.Element = wrapper.find('ul');
      expect(ul).toHaveLength(1);

      wrapper.find('h3').simulate('click');
      ul = wrapper.find('ul');
      expect(ul).toHaveLength(0);
    });

    it('should change title if list element is chosen', () => {
      let title = wrapper.find('h3');
      title.simulate('click');
      let firestTitleText = title.text();
      expect(firestTitleText).toEqual('all genres');

      const element = wrapper.find('li:last-child');
      const elementText = 'american';
      element.simulate('click', {
        target: { innerText: elementText }
      });
      title = wrapper.find('h3');
      const secondTitleText = title.text();
      expect(secondTitleText).toEqual(elementText);
    });

    it('should hide opened unordered list if list element is chosen', () => {
      wrapper.find('h3').simulate('click');
      let ul: JSX.Element = wrapper.find('ul');
      expect(ul).toHaveLength(1);

      const element = wrapper.find('li:last-child');
      element.simulate('click', {
        target: { innerText: 'american' }
      });
      ul = wrapper.find('ul');
      expect(ul).toHaveLength(0);
    });
  });

  describe("adding action creators calls", () => {
    it('should call addStateFilter with chosen state after click', () => {
      const mockState: string = 'CO';
      wrapper = shallow(<Filter type="states" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:last-child').simulate('click', {
        target: { innerText: mockState }
      });
      expect(addStateFilter).toHaveBeenCalledWith(mockState);
    });

    it('should call addGenreFilter with chosen state after click', () => {
      const mockGenre: string = 'american';
      wrapper = shallow(<Filter type="genres" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:last-child').simulate('click', {
        target: { innerText: mockGenre }
      });
      expect(addGenreFilter).toHaveBeenCalledWith(mockGenre);
    });

    it('should call addAttireFilter with chosen state after click', () => {
      const mockAttire: string = 'casual';
      wrapper = shallow(<Filter type="attires" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:last-child').simulate('click', {
        target: { innerText: mockAttire }
      });
      expect(addAttireFilter).toHaveBeenCalledWith(mockAttire);
    });
  });

  describe("removing action creators calls", () => {
    it('should call removeStateFilter after first list element was click', () => {
      wrapper = shallow(<Filter type="states" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:first-child').simulate('click');
      expect(removeStateFilter).toHaveBeenCalled();
    });

    it('should call removeGenreFilter after first list element was click', () => {
      wrapper = shallow(<Filter type="genres" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:first-child').simulate('click');
      expect(removeGenreFilter).toHaveBeenCalled();
    });

    it('should call removeAttireFilter after first list element was click', () => {
      wrapper = shallow(<Filter type="attires" />);
      wrapper.find('h3').simulate('click');

      wrapper.find('li:first-child').simulate('click');
      expect(removeAttireFilter).toHaveBeenCalled();
    });
  });
});
