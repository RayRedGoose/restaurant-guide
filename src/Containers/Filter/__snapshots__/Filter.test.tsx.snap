// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should match the snapshot if type is attires 1`] = `
<section
  className="filter"
>
  <h3
    onClick={[Function]}
  >
    all attires
  </h3>
</section>
`;

exports[`Filter should match the snapshot if type is genres 1`] = `
<section
  className="filter"
>
  <h3
    onClick={[Function]}
  >
    all genres
  </h3>
</section>
`;

exports[`Filter should match the snapshot if type is states 1`] = `
<section
  className="filter"
>
  <h3
    onClick={[Function]}
  >
    all states
  </h3>
</section>
`;
